File: src/lib/totp.ts
import crypto from "crypto";

function base32Decode(s: string): Buffer {
  const abc = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
  const clean = s.toUpperCase().replace(/[^A-Z2-7]/g, "");
  let bits = 0, value = 0;
  const out: number[] = [];
  for (const ch of clean) {
    const idx = abc.indexOf(ch);
    if (idx < 0) continue;
    value = (value << 5) | idx;
    bits += 5;
    if (bits >= 8) { out.push((value >>> (bits - 8)) & 0xff); bits -= 8; }
  }
  return Buffer.from(out);
}

/** Стандартний TOTP: SHA-1, 6 цифр, вікно 30 секунд */
export function totp(secretBase32: string, at = Date.now()): { code: string; secondsLeft: number } {
  const step = 30;
  const counter = Math.floor(at / 1000 / step);
  const buf = Buffer.alloc(8);
  buf.writeBigUInt64BE(BigInt(counter));
  const hmac = crypto.createHmac("sha1", base32Decode(secretBase32)).update(buf).digest();
  const off = hmac[hmac.length - 1] & 0x0f;
  const bin = ((hmac[off] & 0x7f) << 24) | (hmac[off + 1] << 16) | (hmac[off + 2] << 8) | hmac[off + 3];
  return {
    code: String(bin % 1_000_000).padStart(6, "0"),
    secondsLeft: step - Math.floor((at / 1000) % step),
  };
}

File: backend/totp.py
"""RFC 6238 TOTP: SHA-1, 6 цифр, вікно 30 секунд."""
from __future__ import annotations

import hashlib
import hmac
import struct
import time


def _base32_decode(secret: str) -> bytes:
    abc = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
    clean = "".join(ch for ch in (secret or "").upper() if ch in abc)
    bits = 0
    value = 0
    out = bytearray()
    for ch in clean:
        idx = abc.index(ch)
        value = (value << 5) | idx
        bits += 5
        if bits >= 8:
            out.append((value >> (bits - 8)) & 0xFF)
            bits -= 8
    return bytes(out)


def generate(secret_base32: str, at: float | None = None) -> dict[str, int | str]:
    now = at if at is not None else time.time()
    step = 30
    counter = int(now // step)
    msg = struct.pack(">Q", counter)
    digest = hmac.new(_base32_decode(secret_base32), msg, hashlib.sha1).digest()
    offset = digest[-1] & 0x0F
    code_int = struct.unpack(">I", digest[offset : offset + 4])[0] & 0x7FFFFFFF
    return {
        "code": str(code_int % 1_000_000).zfill(6),
        "secondsLeft": step - int(now) % step,
    }
